### DISCUSSION TITLE
Help to parse this file

### DISCUSSION THREAD

**Discussion by joseluisgs** [pdvrieze/xmlutil#103] [Q&A]
Hi
I need to parse this file. I am new user of serialization in XML and I don't know if i need to use a special annotation.
I will appreciate any help
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<response>

<resources name="response">
  <resource>
    <str name="ayto:particles"/>
    <str name="ayto:NO2">116.0</str>
    <str name="ayto:type">AirQualityObserved</str>
    <str name="ayto:latitude">43.449</str>
    <str name="ayto:temperature">22.9</str>
    <str name="ayto:altitude"/>
    <str name="ayto:speed"/>
    <str name="ayto:CO">0.1</str>
    <date name="dc:modified">2021-08-27T07:47:39Z</date>
    <str name="dc:identifier">3047</str>
    <str name="ayto:longitude">-3.8677</str>
    <str name="ayto:odometer"/>
    <str name="ayto:course"/>
    <str name="ayto:ozone">120.0</str>
    <str name="uri">http://datos.santander.es/api/datos/sensores_smart_mobile/3047.xml</str></resource>
  <resource>

 <resource>
    <str name="ayto:particles"/>
    <str name="ayto:NO2">116.0</str>
    <str name="ayto:type">AirQualityObserved</str>
    <str name="ayto:latitude">43.449</str>
    <str name="ayto:temperature">22.9</str>
    <str name="ayto:altitude"/>
    <str name="ayto:speed"/>
    <str name="ayto:CO">0.1</str>
    <date name="dc:modified">2021-08-27T07:47:39Z</date>
    <str name="dc:identifier">3047</str>
    <str name="ayto:longitude">-3.8677</str>
    <str name="ayto:odometer"/>
    <str name="ayto:course"/>
    <str name="ayto:ozone">120.0</str>
    <str name="uri">http://datos.santander.es/api/datos/sensores_smart_mobile/3047.xml</str></resource>
  <resource>

<resources/>
<response>
```
A response class have a item resouces (with name and a list of resouce)
My problem is how to parse str with name and values

Thank you for the help

**Comment by pdvrieze** [comment]
The XML example isn't 100% well-formed, but I assume that it is just an error.

There are some challenges here, including value dependent types. The way to resolve such must include a custom serializer. I would advise
to use a delegate object approach (see [The delegate example MyXmlDelegate.kt](https://github.com/pdvrieze/xmlutil/blob/master/examples/src/main/kotlin/net/devrieze/serialization/examples/customserializer/MyXmlDelegate.kt)). In addition you can use a sealed child class to support different types of tags inside the name/value list.

Then it is up to normal code to translate the list of attributes to a regular resource, and the other way around. Unfortunately formats
like these are too common but against the spirit of XML in some ways. They prevent the parser from doing enforcement (note that the platform independent parser doesn't enforce anyway) and don't work in line with schema's. The serialization format obviously cannot do semantic handling either (it is generic and doesn't know about your stuff) so the custom parser/constructor will need to do it.

The advantage of the chosen approach is that the DTO can remain private (you may need to fiddle a bit with the way it is private) and encapsulated, but going this way is generally easier and more maintainable than actually writing a fully custom serializer (rather than a delegating one).



I would go something like (please note, I didn't try to compile it so there may be some small errors in it):
```kotlin
@Serializable(with=Resource.Companion::class)
@Serialname("resource")
class Resource constructor(/*...*/) {
    private constructor(orig: ResourceDTO)


    @Serializable
    @Serialname("resource")
    private class ResourceDTO(val data: List<ValueDTO>) {
        constructor(orig: Resource): this(TODO("Some way to get the data out of the resource")
    }

    companion object : KSerializer<Resource> {
       override fun deserialize(decoder: Decoder): Resource {
          Resource(ResourceDTO.serializer().deserialize(strategy))
       }

       override fun serializeserialize(encoder: Encoder, value: Resource) {
           ResourceDTO.serializer().serialize(encoder, ResourceDTO(value))
       }
    }
}

@Serializable
private sealed class ValueDTO

@Serializable
@Serialname("str")
private class StrValueDTO(
    @XmlElement(false) val name:String, 
    @XmlValue(true) val value: String
)


@Serializable
@Serialname("date")
private class DateValueDTO(
    @XmlElement(false) val name:String, 
    @XmlValue(true) val value: String // Could be some datetime value with an appropriate serializer (as long as it serializes to some string value not struct).
)
```

**Comment by joseluisgs** [comment]
Oh thank you so much! I would try it.
The file is a XMl about Open data, it is a fragment of a real file. I don't have any problem if I use JDOM and make my own parser, buy I would like to do only with you technology and Kotlin serializer.
I will try it and the I will post the results.
Thank you

**Reply by pdvrieze** [reply]
You're welcome. My intent is for the format to support most reasonable ways of writing xml, but at this point there is no direct support for key/value mapping to attributes. That would be an interesting addition, but in the meantime the delegate approach should allow this to be done reasonably easily.
